lib/srv/discovery/fetchers/eks.go

```diff
@@ -0,0 +1,166 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fetchers
+
+import (
+	"context"
+	"sync"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/eks"
+	"github.com/aws/aws-sdk-go/service/eks/eksiface"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+type eksFetcher struct {
+	EKSFetcherConfig
+}
+
+// EKSFetcherConfig configures the EKS fetcher.
+type EKSFetcherConfig struct {
+	// Client is the AWS eKS client.
+	Client eksiface.EKSAPI
+	// Region is the region where the clusters should be located.
+	Region string
+	// FilterLabels are the filter criteria.
+	FilterLabels types.Labels
+	// Log is the logger.
+	Log logrus.FieldLogger
+}
+
+// CheckAndSetDefaults validates and sets the defaults values.
+func (c *EKSFetcherConfig) CheckAndSetDefaults() error {
+	if c.Client == nil {
+		return trace.BadParameter("missing Client field")
+	}
+	if len(c.Region) == 0 {
+		return trace.BadParameter("missing Region field")
+	}
+
+	if len(c.FilterLabels) == 0 {
+		return trace.BadParameter("missing FilterLabels field")
+	}
+
+	if c.Log == nil {
+		c.Log = logrus.New()
+	}
+	return nil
+}
+
+// NewEKSFetcher creates a new EKS fetcher configuration.
+func NewEKSFetcher(cfg EKSFetcherConfig) (Fetcher, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &eksFetcher{cfg}, nil
+}
+
+func (a *eksFetcher) Get(ctx context.Context) (types.ResourcesWithLabels, error) {
+	clusters, err := a.getEKSClusters(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return clusters.AsResources(), nil
+}
+
+func (a *eksFetcher) getEKSClusters(ctx context.Context) (types.KubeClusters, error) {
+	var (
+		clusters types.KubeClusters
+		mu       sync.Mutex
+	)
+	err := a.Client.ListClustersPagesWithContext(ctx,
+		&eks.ListClustersInput{
+			Include: nil, // For now we should only list EKS clusters
+		},
+		func(clustersList *eks.ListClustersOutput, _ bool) bool {
+			wg := &sync.WaitGroup{}
+			wg.Add(len(clustersList.Clusters))
+			for i := 0; i < len(clustersList.Clusters); i++ {
+				go func(clusterName string) {
```

Comment:
Couple of comments here:

- Can this function be moved out into a separate method to reduce indentation? Also, I would make a function that returns an error and call it here inside goroutine instead of doing "logger.Warn(); return" at each exit point so you can log the error only once.
- This will make the number of API calls equal to the number of clusters. At the very least we should limit the parallelism here (e.g. using `errgroup`), but can this be optimized a bit and e.g. describe clusters not one-by-one by in a batch to account for scenarios when there's a lot of them? E.g. using [ListClusters](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/eks#Client.ListClusters).

Reply:
For the second point, it's not possible to describe clusters in batches. If you look at the [`ListClusters`](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/eks#ListClustersOutput) output it just returns an array of strings with the cluster names available for the region. No other details are present and we need to at least have the tags + current cluster status. These values are only available via `DescribeCluster` method which only accepts one cluster per call. 

We use `ListClustersPagesWithContext`  which is a paginated version of `ListClusters` but the output is exactly the same.

Reply:
Gotcha, that's a bummer but ok.